### PR TITLE
Add support for Git Log API.

### PR DESCRIPTION
**Pull request by ShimonBenYair**
Add support for skip and maxCount when calling the LOG API.
Add support to get the log for a specific folder or file in the repository.
For every commit that is returned when calling the LOG API, the following data was added:
The author of the commit
A list of branch names that the commit is related to.
A list of file names that were changed/deleted/added in the commit.
A list of commit parents (to enable for example clients to paint a graph)

Change-Id: Ibd812784ee5f4208c285e15bf614750ee7e1dd55
Signed-off-by: Shimon Ben.Yair shimon.ben.yair@sap.com

### CODE REVIEW

**Comment by ShimonBenYair** [comment]
@tareqhs @offer8

**Comment by ShimonBenYair** [comment]
@skabashnyuk  @vinokurig  
Hi Do you have any comments ?

File: wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java [hunk 1]
```diff
@@ -752,9 +884,20 @@ private void setRevisionRange(LogCommand logCommand, LogRequest request) throws
...
+                logCommand.setSkip((skip - 1) * maxCount);
+                logCommand.setMaxCount(maxCount);
+            }
+        }
+    }
+
     @Override
     public List<GitUser> getCommiters() throws GitException {
-        List<GitUser> gitUsers = new ArrayList<>();
+        List<GitUser> gitUsers = new ArrayList<GitUser>();
```
Comment (sleshchenko) [inline wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java]: Why did you remove diamond operator?

File: wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java [hunk 2]
```diff
@@ -706,44 +716,166 @@ public void init(InitRequest request) throws GitException {
...
+                                                     .withAuthor(oAuthor)
+                                                     .withBranches(commitBranchesList)
+                                                     .withCommitParent(commitParentsList)
+                                                     .withDiffCommitFile(diffCommitFilesList);
+        return revision;
+    }
+
+    private GitUser getCommitCommitter(RevCommit commit) {
+        PersonIdent committerIdentity = commit.getCommitterIdent();
+        GitUser oCommitter = createDto(GitUser.class).withName(committerIdentity.getName())
```
Comment (sleshchenko) [inline wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java]: Please explain Where is createDto method declared?

File: wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java [hunk 3]
```diff
@@ -706,44 +716,166 @@ public void init(InitRequest request) throws GitException {
...
+                    }
+                }
+            }
+        } finally {
+            if (tw != null) {
+                tw.close();
+            }
+        }
+        if (diffs != null) {
+            for (int i = 0; i < diffs.size(); i++) {
```
Comment (sleshchenko) [inline wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java]: I would rewrite it with java8

```
commitFilesList.addAll(diffs.stream().
                            .map(diff -> DtoFactory.newDto(DiffCommitFile.class)
                                                   .withOldPath(diff.getOldPath())
                                                   .withNewPath(diff.getNewPath())
                                                   .withChangeType(diff.getChangeType().name()))
                            .collect(Collectors.toList()));
```

File: wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java [hunk 4]
```diff
@@ -706,44 +716,166 @@ public void init(InitRequest request) throws GitException {
...
+            throw new GitException(e);
+        }
+    }
+
+    private Revision getRevision(RevCommit commit, String sFilePath) throws GitAPIException, IOException {
+        List<String> commitParentsList = getCommitParents(commit);
+        List<Branch> commitBranchesList = getCommitBranches(commit, ListMode.ALL);
+        List<DiffCommitFile> diffCommitFilesList = getCommitDiffFiles(commit,sFilePath);
+
+        GitUser oAuthor = getCommitAuthor(commit);
```
Comment (sleshchenko) [inline wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java]: Can you explain why is it better to use o prefix in variable name?

File: wsagent/che-core-api-git-shared/src/main/java/org/eclipse/che/api/git/shared/DiffCommitFile.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+ * Copyright (c) 2012-2015 Codenvy, S.A.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Codenvy, S.A. - initial API and implementation
+ *******************************************************************************/
+
```
Comment (sleshchenko) [inline wsagent/che-core-api-git-shared/src/main/java/org/eclipse/che/api/git/shared/DiffCommitFile.java:11]: Please remove empty line here

File: wsagent/che-core-api-git-shared/src/main/java/org/eclipse/che/api/git/shared/DiffCommitFile.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+package org.eclipse.che.api.git.shared;
+
+import org.eclipse.che.dto.shared.DTO;
+
+/**
+ * Created by I048384 on 24/05/2016.
+ */
+
```
Comment (sleshchenko) [inline wsagent/che-core-api-git-shared/src/main/java/org/eclipse/che/api/git/shared/DiffCommitFile.java:19]: Please remove empty line here

File: wsagent/che-core-api-git-shared/src/main/java/org/eclipse/che/api/git/shared/DiffCommitFile.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+ * Contributors:
+ *   Codenvy, S.A. - initial API and implementation
+ *******************************************************************************/
+
+package org.eclipse.che.api.git.shared;
+
+import org.eclipse.che.dto.shared.DTO;
+
+/**
+ * Created by I048384 on 24/05/2016.
```
Comment (sleshchenko) [inline wsagent/che-core-api-git-shared/src/main/java/org/eclipse/che/api/git/shared/DiffCommitFile.java:17]: Please rewrite java docs in following format

```
/**
* @authror Name Surname
*/
```

File: wsagent/che-core-api-git-shared/src/main/java/org/eclipse/che/api/git/shared/Revision.java [hunk 8]
```diff
@@ -20,34 +21,148 @@
...
+
+    /**
+     * Create a Revision object based on a given list of commit parents
+     *
+     * @param commitParents
+     *         a list of commit parents
+     * @return a Revision object
+     */
+    Revision withCommitParent(List<String> commitParents);
 }
```
Comment (vinokurig) [inline wsagent/che-core-api-git-shared/src/main/java/org/eclipse/che/api/git/shared/Revision.java]: add new line

File: wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java [hunk 9]
```diff
@@ -706,44 +716,166 @@ public void init(InitRequest request) throws GitException {
...
+        return oAuthor;
+    }
+
+    private List<Branch>  getCommitBranches(RevCommit commit, ListMode mode) throws GitAPIException {
+        List<Branch> branchList = new ArrayList<Branch>();
+        List<Ref> branches = getGit().branchList()
+                                     .setListMode(mode)
+                                     .setContains(commit.getName())
+                                     .call();
+        for (int i = 0; i < branches.size(); i++) {
```
Comment (sleshchenko) [inline wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java]: Why do you use fori loop instead of foreach? BTW I would use stream API here

File: wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java [hunk 10]
```diff
@@ -706,44 +716,166 @@ public void init(InitRequest request) throws GitException {
...
-                                                          .withCommitTime(MILLISECONDS.convert(commit.getCommitTime(), SECONDS))
-                                                          .withCommitter(gitUser);
+                Revision revision = getRevision(commit, sFilePath);
                 commits.add(revision);
             }
             return new LogPage(commits);
-        } catch (GitAPIException | IOException exception) {
-            String errorMessage = exception.getMessage();
-            if (ERROR_LOG_NO_HEAD_EXISTS.equals(errorMessage)) {
-                throw new GitException(errorMessage, ErrorCodes.INIT_COMMIT_WAS_NOT_PERFORMED);
```
Comment (vinokurig) [inline wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java:737]: We need exception with this error code for our client.

File: wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java [hunk 11]
```diff
@@ -706,44 +716,166 @@ public void init(InitRequest request) throws GitException {
...
+        for (int i = 0; i < branches.size(); i++) {
+            branchList.add(createDto(Branch.class).withName(branches.get(i).getName()));
+        }
+        return branchList;
+    }
+
+    private List<String> getCommitParents(RevCommit commit) {
+        RevCommit[] parents = commit.getParents();
+        List<String> commitParentsList = new ArrayList<String>();
+        for (int i=0;i<parents.length;i++) {
```
Comment (vinokurig) [inline wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java]: Please format this line (add spaces)

File: wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java [hunk 12]
```diff
@@ -706,44 +716,166 @@ public void init(InitRequest request) throws GitException {
...
-            if (ERROR_LOG_NO_HEAD_EXISTS.equals(errorMessage)) {
-                throw new GitException(errorMessage, ErrorCodes.INIT_COMMIT_WAS_NOT_PERFORMED);
+        } catch (GitAPIException | IOException e) {
+            LOG.error("Failed to retrieve log. ", e);
+            throw new GitException(e);
+        }
+    }
+
+    private Revision getRevision(RevCommit commit, String sFilePath) throws GitAPIException, IOException {
+        List<String> commitParentsList = getCommitParents(commit);
```
Comment (sleshchenko) [inline wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java]: I think you can rewrite it as 

```
Strean.of(commit.getParents())
       .map(RevCommit::getName)
       .collect(toList());
```

and remove private method

**Comment by ShimonBenYair** [comment]
@skabashnyuk  @vinokurig 

We had some technical issues, so we closed this pull request
and created a new one with all the fixes you mentioned.
The new branch is : add_support_git_log_API
